### PR TITLE
/users/search first and last name added in search

### LINKED ISSUE #1882: /users?search API only works for username and not for firstName & lastName
### Issue Description
The `/users?search=<value>` API only works for usernames, meaning it searches exclusively for usernames and not for first names or last names. Consequently, it can sometimes be difficult to determine a person's username.

### Expected Behavior
`/users?search=rishi should search for usernames, first names, and last names.

### Current Behavior
`/users?search=rishi only queries the username field, even when it perfectly matches the first name or last name.

### Proof

https://github.com/Real-Dev-Squad/website-backend/assets/56365512/e7b766ad-f9bd-4c43-829d-f4273daad8f8

### Reproducibility

 - [x] This issue is reproducible
 - [ ] This issue is not reproducible

### Steps to Reproduce
1.  Go to `https://staging-status.realdevsquad.com/tasks`
2. Hold option/alt key
3. Search for user

### Severity/Priority
 - [ ] Critical
 - [ ] High
 - [x] Medium
 - [ ] Low

### Additional Information
NA

### Checklist

- [x] I have read and followed the project's code of conduct.
- [x] I have searched for similar issues before creating this one.
- [x] I have provided all the necessary information to understand and reproduce the issue.
- [ ] I am willing to contribute to the resolution of this issue.

### PR DESCRIPTION
**Pull request by ashifkhn**
Date: 15th March 2024

Developer Name: Ashif

---

## Issue Ticket Number

closes #1882

## Description
I have wrote a new query to fetch users by first,last and username. In this change now we can get the user to assign a task by the first,last and username instead of just username.

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

Screenshot 2024-03-15 at 17 43 30

</details>

## Test Coverage
<details>
<summary>Screenshot 1</summary>

</details>

## Additional Notes
I'll raise test in next pr.I'm raising this one to get reviewed

### CODE REVIEW

**Review by Achintya-Chatterjee — DISMISSED** [review]

File: models/users.js [hunk 1]
```diff
@@ -163,57 +163,155 @@ const getSuggestedUsers = async (skill) => {
...
+  if (isDevMode) {
+    try {
+      const size = parseInt(query.size) || 100;
+      const doc = (query.next || query.prev) && (await userModel.doc(query.next || query.prev).get());
+
+      let dbQuery = userModel.where("roles.archived", "==", false).orderBy("username");
+      const usernameQuery = userModel.where("roles.archived", "==", false).orderBy("username");
+      const firstNameQuery = userModel.where("roles.archived", "==", false).orderBy("first_name");
+      const lastNameQuery = userModel.where("roles.archived", "==", false).orderBy("last_name");
+      let compositeQuery = [usernameQuery, firstNameQuery, lastNameQuery];
```
Comment (Achintya-Chatterjee) [inline models/users.js:177]: `fetchPaginatedUsers` can be refactored to separate concerns into distinct functions or modules, such as query construction, data retrieval, and error handling. composite query construction can be abstracted into a reusable function to eliminate redundancy and to improve maintainability.

**Review by Achintya-Chatterjee — CHANGES_REQUESTED** [review]

File: models/users.js [hunk 2]
```diff
@@ -163,57 +163,155 @@ const getSuggestedUsers = async (skill) => {
...
+        allUsers,
+        nextId: lastDoc?.id ?? "",
+        prevId: firstDoc?.id ?? "",
+      };
+    } catch (err) {
+      logger.error("Error retrieving user data", err);
+      throw err;
+    }
+  } else {
+    try {
```
Comment (ardourApeX) [inline models/users.js:263]: why do we have 2 try catch block just because of the if-else condition?

File: models/users.js (hunk 1 above)
Comment (ashifkhn) [reply models/users.js]: Please elaborate.

Comment (ashifkhn) [reply models/users.js]: is this a p* for this task?

**Review by Achintya-Chatterjee — APPROVED** [review]

File: models/users.js [hunk 3]
```diff
@@ -163,53 +163,74 @@ const getSuggestedUsers = async (skill) => {
...
+    snapshots.forEach(processSnapshot);
+
+    const firstDoc = snapshot.docs[0];
+    const lastDoc = snapshot.docs[snapshot.docs.length - 1];
 
     return {
       allUsers,
-      nextId: lastDoc?.id ?? "",
-      prevId: firstDoc?.id ?? "",
+      nextId: lastDoc?.id || "",
```
Comment (shreya-mishra) [inline models/users.js:232]: can nextID be empty string? If yes then why?

File: models/users.js [hunk 4]
```diff
@@ -163,53 +163,74 @@ const getSuggestedUsers = async (skill) => {
...
+      const lastNameQuery = userModel.where("roles.archived", "==", false).orderBy("last_name");
+      compositeQuery = [usernameQuery, firstNameQuery, lastNameQuery];
+    }
 
     if (query.prev) {
+      compositeQuery = compositeQuery.map((query) => query.limitToLast(size));
       dbQuery = dbQuery.limitToLast(size);
     } else {
+      compositeQuery = compositeQuery.map((query) => query.limit(size));
       dbQuery = dbQuery.limit(size);
```
Comment (ardourApeX) [inline models/users.js:186]: As dbQuery will itself lie inside composite query, why do we need to use it here as well?

File: models/users.js [hunk 5]
```diff
@@ -163,53 +163,74 @@ const getSuggestedUsers = async (skill) => {
...
       } else if (query.prev) {
         dbQuery = dbQuery.endBefore(doc);
+        compositeQuery = compositeQuery.map((query) => query.endBefore(doc));
       }
     }
-    const snapshot = await dbQuery.get();
 
-    const firstDoc = snapshot.docs[0];
-    const lastDoc = snapshot.docs[snapshot.docs.length - 1];
+    const snapshot = await dbQuery.get();
```
Comment (ardourApeX) [inline models/users.js:208]: when the dev flag is turned off, snapshots will already contain dbQuery, so no need to retrieve it's corresponding data separately.

**Review by ardourApeX — CHANGES_REQUESTED** [review]

File: test/integration/users.test.js [hunk 6]
```diff
@@ -1246,6 +1246,42 @@ describe("Users", function () {
...
+        .query({ search: "Ankur", dev: true }) // Search for users with first name 'Ankur' in dev mode
+        .set("cookie", `${cookieName}=${jwt}`)
+        .end((err, res) => {
+          if (err) {
+            return done(err);
+          }
+          expect(res).to.have.status(200);
+          expect(res.body).to.be.a("object");
+          expect(res.body.message).to.equal("Users returned successfully!");
+          expect(res.body.users).to.be.a("array");
```
Comment (ardourApeX) [inline test/integration/users.test.js:1262]: Please least a single test that denotes that it is now searching on all 3 fields i.e. firstName, lastName, username.

**Review by Achintya-Chatterjee — APPROVED** [review]

**Review by iamitprakash — APPROVED** [review]

**Pull request merged**